geopandas/io/parquet.py

```diff
@@ -0,0 +1,285 @@
+    ----------
+    path : str
+        File path or Root Directory path. Will be used as Root Directory path
+        while writing a partitioned dataset.
+    compression : {'snappy', 'gzip', 'brotli', None}, default 'snappy'
+        Name of the compression to use. Use ``None`` for no compression.
+    index : bool, default None
+        If ``True``, include the dataframe's index(es) in the file output. If
+        ``False``, they will not be written to the file. If ``None``, the
+        engine's default behavior will be used.
+    kwargs
+        Additional keyword arguments passed to the engine
+    """
+
+    import_optional_dependency(
+        "pyarrow.parquet", extra="pyarrow is required for parquet support."
+    )
+    from pyarrow import parquet, Table
+
+    _validate_dataframe(df)
+
+    geometry_column = df._geometry_column_name
+    geometry_columns = df.columns[df.dtypes == "geometry"]
+
+    # Construct metadata for each geometry
+    column_metadata = {}
+    for col in geometry_columns:
+        series = df[col]
+        column_metadata[col] = {
+            "crs": series.crs.to_wkt() if series.crs else None,
+            "encoding": "WKB",
+            "bounds": series.total_bounds.tolist(),
+        }
+
+    # Convert to a DataFrame so we can convert geometries to WKB while
+    # retaining original column names.
+    df = DataFrame(df.copy())
+
+    # Encode all geometry columns to WKB
+    for col in geometry_columns:
+        df[col] = to_wkb(df[col].values)
+
+    table = Table.from_pandas(df, preserve_index=index)
+
+    # Store geopandas specific file-level metadata
+    # This must be done AFTER creating the table or it is not persisted
+    metadata = table.schema.metadata
+    metadata.update(
+        {
+            "geo": _encode_metadata(
+                {
+                    "primary_column": geometry_column,
+                    "columns": column_metadata,
+                    "schema_version": METADATA_VERSION,
+                    "creator": {
+                        "library": "geopandas",
+                        "version": geopandas.__version__,
+                    },
+                }
+            )
+        }
+    )
+
+    table = table.replace_schema_metadata(metadata)
+    parquet.write_table(table, path, compression=compression, **kwargs)
+
+
+def read_parquet(path, columns=None, **kwargs):
+    """
+    Load a parquet object from the file path, returning a GeoDataFrame.
+
+    You can read a subset of columns in the file using the ``columns`` parameter.
+    However, the structure of the returned GeoDataFrame will depend on which
+    columns you read:
+    * if no geometry columns are read, this will raise a ``ValueError`` - you
+      should use the pandas `read_parquet` method instead.
+    * if the primary geometry column saved to this file is not included in
+      columns, the first available geometry column will be set as the geometry
+      column of the returned GeoDataFrame.
```

Comment:
I think forcing a read of `primary_column` runs counter to allowing subsets of columns to be read, and allowing multiple geometry columns to exist in the dataset.  We may want to serialize 3 geometry columns, but read only one of them, which was not the original primary.  `primary_column` is a hint to `geopandas` that a given column was the primary one when written.

We could allow the user to pass in the name of the column they want to be primary after read, but I just ripped a bunch of that out because it presents a number of edge cases that need to be validated.

Reply:
> We may want to serialize 3 geometry columns, but read only one of them, which was not the original primary

Hmm, that's a good point.

Another option would be to, for now, only automatically set a different column as primary column, if there is only one geometry column in the selection, or raise a warning when there are multiple and we are guessing.

I suppose for this use case, adding a keyword to select the primary column, is maybe the best after all. But I am fine with leaving that for a separate PR if that needs additional testing of edge cases

(we should start listing those follow-up items somewhere)

Reply:
Will raise warning in this PR.  Adding parameter noted in #1382